[test] failures due to error in gameplay framework

Relevant error:
```        
if messages and messages.messages:
          for msg in messages.messages:
              msg_obj = Message(
                  sender=self.power_name,
                  recipient=msg['recipient'],
                  message=msg['message'],
                  phase=self.game.get_current_phase(),
              )
>               yield game.send_game_message(message=msg_obj)
E               NameError: name 'game' is not defined
```

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -25,6 +25,6 @@ jobs:
     - name: "Run tests"
       run: |
         pytest tests/utils_test.py
-        # pytest tests/bot_tests/bots_test.py
+        pytest tests/bot_tests/bots_test.py
         # pytest tests/randomize_order_test.py
         # pytest tests/bot_tests/soa_bot_functions_test.py

--- a/docs/source/containerization.rst
+++ b/docs/source/containerization.rst
@@ -13,22 +13,22 @@ Building the container
 
 .. code-block:: python
 
-cd containers/allan_dip_bot/
+	cd containers/allan_dip_bot/
 
 2. Download and unzip the model file:
 
 .. code-block:: bash
 
-$ wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
-$ mkdir bot_neurips2019-sl_model
-$ unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
+	$ wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
+	$ mkdir bot_neurips2019-sl_model
+	$ unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
 
 3. Build the docker container with the tag to be used while pushing it to DockerHub:
 
 .. code-block:: bash
 
-$ docker build -t uname/allan_dip_bot .
-$ docker push uname/allan_dip_bot
+	$ docker build -t uname/allan_dip_bot .
+	$ docker push uname/allan_dip_bot
 
 Note, here ``uname`` is supposed to be your username on DockerHub. This is needed if you plan to push the image to DockerHub. If you intend to just create the image and run it locally, you can just write the tag name as ``allan_dip_bot``.
 
@@ -37,14 +37,14 @@ Usage
 
 .. code-block:: bash
 
-$ docker run -it allan_dip_bot --help
---host 		HOST [default localhost]
---port 		PORT [default 8432]
---game_id 	GAME_ID
---power		POWER
---bot_type  BOT_TYPE [default TransparentBot]
---outdir    OUT_DIR
+	$ docker run -it allan_dip_bot --help
+	--host 		HOST [default localhost]
+	--port 		PORT [default 8432]
+	--game_id 	GAME_ID
+	--power		POWER
+	--bot_type  BOT_TYPE [default TransparentBot]
+	--outdir    OUT_DIR
 
-#connect to remote game engine
-$ docker run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
-$ singularity run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
+	#connect to remote game engine
+	$ docker run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
+	$ singularity run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY

--- a/docs/source/containerization.rst
+++ b/docs/source/containerization.rst
@@ -13,36 +13,38 @@ Building the container
 
 .. code-block:: python
 
-    cd containers/allan_dip_bot/
+cd containers/allan_dip_bot/
 
 2. Download and unzip the model file:
 
 .. code-block:: bash
 
-    $ wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
-	$ mkdir bot_neurips2019-sl_model
-	$ unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
+$ wget https://f002.backblazeb2.com/file/ppaquette-public/benchmarks/neurips2019-sl_model.zip
+$ mkdir bot_neurips2019-sl_model
+$ unzip neurips2019-sl_model.zip -d bot_neurips2019-sl_model/
 
 3. Build the docker container with the tag to be used while pushing it to DockerHub:
 
 .. code-block:: bash
 
-    $ docker build -t uname/allan_dip_bot .
-	$ docker push uname/allan_dip_bot
+$ docker build -t uname/allan_dip_bot .
+$ docker push uname/allan_dip_bot
+
+Note, here ``uname`` is supposed to be your username on DockerHub. This is needed if you plan to push the image to DockerHub. If you intend to just create the image and run it locally, you can just write the tag name as ``allan_dip_bot``.
 
 Usage
 ************************************************
 
 .. code-block:: bash
 
-    $ docker run -it allan_dip_bot --help
-	--host 		HOST [default localhost]
-	--port 		PORT [default 8432]
-	--game_id 	GAME_ID
-	--power		POWER
-	--bot_type  BOT_TYPE [default TransparentBot]
-	--outdir    OUT_DIR
+$ docker run -it allan_dip_bot --help
+--host 		HOST [default localhost]
+--port 		PORT [default 8432]
+--game_id 	GAME_ID
+--power		POWER
+--bot_type  BOT_TYPE [default TransparentBot]
+--outdir    OUT_DIR
 
-	#connect to remote game engine
-	$ docker run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
-	$ singularity run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
+#connect to remote game engine
+$ docker run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY
+$ singularity run uname/allan_dip_bot --game_id test_game --host shade.tacc.utexas.edu --power TURKEY

--- a/src/baseline_bots/bots/random_proposer_bot.py
+++ b/src/baseline_bots/bots/random_proposer_bot.py
@@ -101,9 +101,9 @@ class RandomProposerBot_AsyncBot(RandomProposerBot):
                     sender=self.power_name,
                     recipient=msg['recipient'],
                     message=msg['message'],
-                    phase=game.get_current_phase(),
+                    phase=self.game.get_current_phase(),
                 )
-                yield game.send_game_message(message=msg_obj)
+                yield self.game.send_game_message(message=msg_obj)
         orders = yield self.gen_orders()
         # maintain current orders
         self.orders = orders

--- a/src/baseline_bots/bots/smart_order_accepter_bot.py
+++ b/src/baseline_bots/bots/smart_order_accepter_bot.py
@@ -65,7 +65,13 @@ class SmartOrderAccepterBot(DipnetBot):
         self.foes = []
         self.neutral = []
         
-    async def send_message(self, recipient, message):
+    async def send_message(self, recipient: str, message: MessagesData) -> None:
+        """
+        Send message asynchronously to the server while the bot is still processing
+
+        :param recipient: The name of the recipient power
+        :param message: MessagesData object containing set of all messages
+        """
         msg_obj = Message(
             sender=self.power_name,
             recipient=recipient,
@@ -80,6 +86,9 @@ class SmartOrderAccepterBot(DipnetBot):
         """
         Add messages to be sent to powers with positive stance.
         These messages would contain factual information about the orders that current power would execute in current round
+
+        :param msgs_data: MessagesData object containing set of all messages
+        :param orders_list: set of orders that are going to be executed by the bot
         """
         if orders_list:
             orders_decided = FCT(
@@ -98,8 +107,14 @@ class SmartOrderAccepterBot(DipnetBot):
                         msgs_data.add_message(pow, str(orders_decided))
                         await self.send_message(pow, str(orders_decided))
 
-    async def gen_messages(self, orders_list: List[str], msgs_data: MessagesData):
+    async def gen_messages(self, orders_list: List[str], msgs_data: MessagesData) -> MessagesData:
+        """
+        This generates messages to be sent to the other powers. 
+        Note: Messages are also generated outside of this function invocation flow
 
+        :param orders_list: final list of orders decided by the bot
+        :param msgs_data: MessagesData object containing set of all messages
+        """
         # generate messages: we should  be sending our true orders to allies (positive stance)
         await self.gen_pos_stance_messages(msgs_data, orders_list)
 
@@ -110,6 +125,11 @@ class SmartOrderAccepterBot(DipnetBot):
     ) -> MessagesData:
         """
         Reply back to allies regarding their proposals whether we follow or not follow
+
+        :param best_proposer: the name of the best proposer as determined by the bot in utils.get_best_orders
+        :param prp_orders: dictionary of proposed orders
+        :param messages: MessagesData object containing set of all messages
+        :return: MessagesData object containing set of all messages
         """
         for proposer, orders in prp_orders.items():
             if (
@@ -199,32 +219,20 @@ class SmartOrderAccepterBot(DipnetBot):
             print(self.alliances)
 
     def is_support_for_selected_orders(self, support_order: str) -> bool:
-        """Determine if selected support order for neighbour corresponds to a self order selected"""
+        """
+        Determine if selected support order for neighbour corresponds to a self order selected
+
+        :param support_order: the support order to be determined for correspondance with self orders
+        :return: boolean indicating the above mentioned detail
+        """
         order_tokens = get_order_tokens(support_order)
+
+        # Fetch our order for which the support order is determined using supported province name
         selected_order = get_order_tokens(
             self.orders.orders[order_tokens[2].split()[1]]
         )
 
-        if (
-            len(order_tokens[2:]) == len(selected_order)
-            and order_tokens[2:] == selected_order
-        ):
-            # Attack move
-            return True
-        elif selected_order[1].strip() == "H" and (
-            len(order_tokens[2:]) == len(selected_order) - 1
-        ):
-            # Hold move
-            return True
-        return False
-
-    def is_support_for_given_orders(self, support_order, orders):
-        """Determine if selected support order for neighbour corresponds to given list of orders"""
-        order_tokens = get_order_tokens(support_order)
-        if order_tokens[2].split()[1] not in orders:
-            return True  # it's okay to support other power than allies
-        selected_order = get_order_tokens(orders[order_tokens[2].split()[1]])
-
+        # Check if the support order is in correspondance with the order we have selected for our province
         if (
             len(order_tokens[2:]) == len(selected_order)
             and order_tokens[2:] == selected_order
@@ -241,6 +249,8 @@ class SmartOrderAccepterBot(DipnetBot):
     def get_2_neigh_provinces(self) -> Set[str]:
         """
         Determine set of orderable locations of allies which are 1-hop/2-hops away from the current power's orderable locations
+
+        :return: set of orderable locations of neighbouring allies
         """
         provs = [
             loc.upper() for loc in self.game.get_orderable_locations(self.power_name)
@@ -282,7 +292,12 @@ class SmartOrderAccepterBot(DipnetBot):
         return n2n_provs
 
     def bad_move(self, order: str) -> bool:
-        """If order indicates attack on one of its provinces, return True, else return False"""
+        """
+        If order indicates attack on one of its provinces, return True, else return False
+
+        :param order: the order representing the move which is to be determined if it is bad or not
+        :return: boolean indicating the above detail
+        """
         order_tokens = get_order_tokens(order)
 
         if len(order_tokens) == 2:
@@ -297,7 +312,11 @@ class SmartOrderAccepterBot(DipnetBot):
         return False
 
     def support_move(self, order: str) -> bool:
-        """Indicates if order is a support order and is not attacking on one of its provinces"""
+        """
+        Indicates if order is a support order and is not attacking on one of its provinces
+
+        :param order: the order which is to be determined if it is a support move or not
+        """
         order_tokens = get_order_tokens(order)
         if (
             3 <= len(order_tokens) <= 4
@@ -319,7 +338,11 @@ class SmartOrderAccepterBot(DipnetBot):
             self.allies_influence.update(set(self.game.get_power(pow).influence))
 
     def get_allies_orderable_locs(self) -> Set[str]:
-        """Gets provinces which are orderable for the allies"""
+        """
+        Gets provinces which are orderable for the allies
+
+        :return: set of provinces which are orderable for the allies
+        """
         provinces = set()
         for ally in [
             pow1

--- a/tests/bot_tests/SOA_test.py
+++ b/tests/bot_tests/SOA_test.py
@@ -1,5 +1,5 @@
 """unit tests for smart order accepter bot"""
-from gameplay_framework import GamePlayAsync
+from gameplay_framework_async import GamePlayAsync
 from diplomacy import Game, Message
 from tornado import gen
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop

--- a/tests/bot_tests/SOA_test.py
+++ b/tests/bot_tests/SOA_test.py
@@ -1,5 +1,5 @@
 """unit tests for smart order accepter bot"""
-from gameplay_framework import GamePlay
+from gameplay_framework import GamePlayAsync
 from diplomacy import Game, Message
 from tornado import gen
 from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
@@ -83,9 +83,9 @@ class TestSOABot():
         game = Game()
         soa_bot1 = SmartOrderAccepterBot('FRANCE', game)
         soa_bot2 = SmartOrderAccepterBot('RUSSIA', game)
-        game_play = GamePlay(game, [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), soa_bot1, soa_bot2, RandomProposerBot_AsyncBot('GERMANY', game), RandomProposerBot_AsyncBot('ITALY', game), RandomProposerBot_AsyncBot('TURKEY', game)], 3, True)
+        game_play = GamePlayAsync(game, [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), soa_bot1, soa_bot2, RandomProposerBot_AsyncBot('GERMANY', game), RandomProposerBot_AsyncBot('ITALY', game), RandomProposerBot_AsyncBot('TURKEY', game)], 3, True)
         msgs, done = yield game_play.step()
-        game_play = GamePlay(game, [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), soa_bot1, soa_bot2, RandomProposerBot_AsyncBot('GERMANY', game), RandomProposerBot_AsyncBot('ITALY', game), RandomProposerBot_AsyncBot('TURKEY', game)], 3, True)
+        game_play = GamePlayAsync(game, [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), soa_bot1, soa_bot2, RandomProposerBot_AsyncBot('GERMANY', game), RandomProposerBot_AsyncBot('ITALY', game), RandomProposerBot_AsyncBot('TURKEY', game)], 3, True)
         while not game_play.game.is_game_done:
             msgs, done = yield game_play.step()
         print('finish test_play')
@@ -97,7 +97,7 @@ class TestSOABot():
         game = Game()
         soa_bot = SmartOrderAccepterBot('FRANCE', game)
         bot_instances = [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), RandomProposerBot_AsyncBot('GERMANY', game), soa_bot]
-        game_play = GamePlay(game, bot_instances, 3, True)
+        game_play = GamePlayAsync(game, bot_instances, 3, True)
         game_play.game.set_centers('AUSTRIA', ['VIE','TRI','BUD'], reset=True)
         game_play.game.set_centers('ENGLAND', ['LON'], reset=True)
         game_play.game.set_centers('GERMANY', ['MUN', 'KIE', 'BER','BEL'])
@@ -120,7 +120,7 @@ class TestSOABot():
         game = Game()
         soa_bot = SmartOrderAccepterBot('FRANCE', game)
         bot_instances = [RandomProposerBot_AsyncBot('ENGLAND', game), RandomProposerBot_AsyncBot('GERMANY', game), soa_bot]
-        game_play = GamePlay(game, bot_instances, 3, True)
+        game_play = GamePlayAsync(game, bot_instances, 3, True)
         game_play.game.set_orders('FRANCE', ['A MAR H', 'A PAR H', 'F BRE - PIC'])
         game_play.game.set_orders('ENGLAND', ['A LVP - WAL', 'F EDI - NTH', 'F LON - ENG'])
         game_play.game.set_orders('GERMANY', ['A BER - MUN', 'A MUN - BUR', 'F KIE - HOL'])
@@ -158,7 +158,7 @@ class TestSOABot():
         soa_bot = SmartOrderAccepterBot('FRANCE', game)
         soa_bot.ally_threshold = 1.0
         bot_instances = [RandomProposerBot_AsyncBot('ENGLAND', game), RandomProposerBot_AsyncBot('GERMANY', game), soa_bot]
-        game_play = GamePlay(game, bot_instances, 3, True)
+        game_play = GamePlayAsync(game, bot_instances, 3, True)
         game_play.game.set_centers('ENGLAND', ['LON'], reset=True)
         game_play.game.set_centers('GERMANY', ['MUN', 'KIE', 'BER','BEL'])
         game_play.game.set_centers(soa_bot.power_name, ['PAR','BRE', 'MAR'])
@@ -188,7 +188,7 @@ class TestSOABot():
         baseline1 = RandomProposerBot_AsyncBot('AUSTRIA', game)
         baseline2 = RandomProposerBot_AsyncBot('ENGLAND', game)
         bot_instances = [baseline1, baseline2, soa_bot]   
-        game_play = GamePlay(game, bot_instances, 3, True)
+        game_play = GamePlayAsync(game, bot_instances, 3, True)
         rcvd_messages = game_play.game.filter_messages(messages=game_play.game.messages, game_role='AUSTRIA')
         bl1_msg = yield baseline1.gen_messages(rcvd_messages)
         rcvd_messages = game_play.game.filter_messages(messages=game_play.game.messages, game_role='ENGLAND')
@@ -312,7 +312,7 @@ class TestSOABot():
         game = Game()
         soa_bot = SmartOrderAccepterBot('FRANCE', game)
         bot_instances = [RandomProposerBot_AsyncBot('AUSTRIA', game), RandomProposerBot_AsyncBot('ENGLAND', game), RandomProposerBot_AsyncBot('GERMANY', game), soa_bot]
-        game_play = GamePlay(game, bot_instances, 3, True)
+        game_play = GamePlayAsync(game, bot_instances, 3, True)
         game_play.game.set_centers('AUSTRIA', ['VIE','TRI','BUD'], reset=True)
         game_play.game.set_centers('ENGLAND', ['LON'], reset=True)
         game_play.game.set_centers('GERMANY', ['MUN', 'KIE', 'BER','BEL'])

--- a/tests/bot_tests/bots_test.py
+++ b/tests/bot_tests/bots_test.py
@@ -13,7 +13,6 @@ class TestRPBot():
     def test(self):
         game_play = GamePlay(None, [RandomProposerBot, RandomProposerBot], 3, True)
         actions, done = game_play.step()
-        print(actions)
         game_play = GamePlay(None, [LoyalBot, RandomProposerBot], 3)
         game_play.play()
         # game_play = GamePlay(None, [RealPolitik, RandomProposerBot], 3)

--- a/tests/bot_tests/bots_test.py
+++ b/tests/bot_tests/bots_test.py
@@ -2,7 +2,7 @@
 from gameplay_framework import GamePlay
 
 from baseline_bots.bots.baseline_bot import BaselineBot
-from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
+from baseline_bots.bots.random_proposer_bot import RandomProposerBot
 from baseline_bots.bots.pushover_bot import PushoverBot
 from baseline_bots.bots.random_allier_proposer_bot import RandomAllierProposerBot
 from baseline_bots.bots.random_honest_bot import RandomHonestBot
@@ -11,10 +11,10 @@ from baseline_bots.bots.loyal_bot import LoyalBot
 
 class TestRPBot():
     def test(self):
-        game_play = GamePlay(None, [RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot], 3, True)
+        game_play = GamePlay(None, [RandomProposerBot, RandomProposerBot], 3, True)
         actions, done = game_play.step()
         print(actions)
-        game_play = GamePlay(None, [LoyalBot, RandomProposerBot_AsyncBot], 3)
+        game_play = GamePlay(None, [LoyalBot, RandomProposerBot], 3)
         game_play.play()
         # game_play = GamePlay(None, [RealPolitik, RandomProposerBot], 3)
         # game_play.play()

--- a/tests/bot_tests/gameplay_framework.py
+++ b/tests/bot_tests/gameplay_framework.py
@@ -53,12 +53,11 @@ class GamePlay():
             if type(bot) == BaselineMsgRoundBot:
                 bot.phase_init()
 
-    @gen.coroutine     
     def step(self):
         """one step of messaging"""
 
         if self.game.is_game_done:
-            return True
+            return None, True
         
         # if message rounds over
         if self.cur_local_message_round == self.msg_rounds:
@@ -66,7 +65,7 @@ class GamePlay():
         while self.game.get_current_phase()[-1] != 'M':
             self.game.process()
             if self.game.is_game_done:
-                return True
+                return None, True
 
         round_msgs = self.game.messages
         msgs_to_send = {}

--- a/tests/bot_tests/gameplay_framework.py
+++ b/tests/bot_tests/gameplay_framework.py
@@ -39,12 +39,11 @@ class GamePlay():
         self.cur_local_message_round = 0
         self.phase_init_bots()
 
-    @gen.coroutine  
     def play(self):
         """play a game with the bots"""
 
         while not self.game.is_game_done:
-            yield self.step()
+            self.step()
 
         if self.save_json:
             to_saved_game_format(self.game, output_path='GamePlayFramework.json')
@@ -56,6 +55,74 @@ class GamePlay():
             if type(bot) == BaselineMsgRoundBot:
                 bot.phase_init()
 
+    @gen.coroutine     
+    def step(self):
+        """one step of messaging"""
+
+        if self.game.is_game_done:
+            return True
+        
+        # if message rounds over
+        if self.cur_local_message_round == self.msg_rounds:
+           self.phase_init_bots()
+        while self.game.get_current_phase()[-1] != 'M':
+            self.game.process()
+            if self.game.is_game_done:
+                return True
+
+        round_msgs = self.game.messages
+        msgs_to_send = {}
+        for bot in self.bots:
+            # retrieve messages sent to bot
+            rcvd_messages = self.game.filter_messages(messages=round_msgs, game_role=bot.power_name)
+            
+            # an array of Message objects
+            rcvd_messages = list(rcvd_messages.items())
+            
+            # get messages to be sent from bot
+            ret_dict = bot(rcvd_messages)
+
+
+            if "messages" in ret_dict:
+                bot_messages = ret_dict["messages"]#bot.gen_messages(rcvd_messages)
+                
+                msgs_to_send[bot.power_name] = bot_messages
+
+        # Send all messages after all bots decide
+        for power_name in msgs_to_send:
+            msgs = msgs_to_send[power_name]
+            for msg in msgs:
+                msg_obj = Message(
+                    sender=power_name,
+                    recipient=msg['recipient'],
+                    message=msg['message'],
+                    phase=self.game.get_current_phase(),
+                )
+                self.game.add_message(message=msg_obj)
+
+        # get/set orders
+        for bot in self.bots:
+            if hasattr(bot, "orders"):
+                orders = ret_dict["orders"]
+                if orders is not None:
+                    self.game.set_orders(power_name=bot.power_name, orders=orders)
+
+        self.cur_local_message_round+=1
+
+        self.game.process()
+        return {"messages": msgs_to_send}, self.game.is_game_done
+
+class GamePlayAsync(GamePlay):
+    @gen.coroutine  
+    def play(self):
+        """play a game with the bots"""
+
+        while not self.game.is_game_done:
+            yield self.step()
+
+        if self.save_json:
+            to_saved_game_format(self.game, output_path='GamePlayFramework.json')
+        
     @gen.coroutine     
     def step(self):
         """one step of messaging"""
@@ -112,11 +179,11 @@ class GamePlay():
 
         self.game.process()
         return {"messages": msgs_to_send}, self.game.is_game_done
-            
+
 
 @gen.coroutine            
 def game_loop():
-    game_play_obj = GamePlay(None, [RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot], 3, True)
+    game_play_obj = GamePlayAsync(None, [RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot], 3, True)
     yield game_play_obj.play()
     stop_io_loop()
                 

--- a/tests/bot_tests/gameplay_framework_async.py
+++ b/tests/bot_tests/gameplay_framework_async.py
@@ -7,52 +7,25 @@ from diplomacy import Game, Message, connect
 from diplomacy.utils.export import to_saved_game_format
 
 from baseline_bots.bots.baseline_bot import BaselineBot, BaselineMsgRoundBot
+from baseline_bots.bots.random_proposer_bot import RandomProposerBot_AsyncBot
+from diplomacy_research.utils.cluster import start_io_loop, stop_io_loop
 import sys
 sys.path.append("../../../dipnet_press")
 
-class GamePlay():
-    """
-    A simple framework to test multiple bots together
-    """
-    def __init__(self, game:Game, bots:List[BaselineBot], msg_rounds:int, save_json=False):
-        assert(len(bots) <= 7), "too many bots"
-        # if no game is passed, assume bots is a list of bot classes to
-        # be instantiated. 
-        if game is None:
-            # make game
-            game = Game()
-            # get list of powers
-            power_names = list(game.get_map_power_names())
-            inst_bots = []
-            # instantiate each bot
-            for i, bot_class in enumerate(bots):
-                inst_bots.append(bot_class(power_names[i], game))
-            self.bots = inst_bots
-        else:
-            self.bots = bots 
+from gameplay_framework import GamePlay
 
-        self.game = game
-        self.msg_rounds = msg_rounds
-        self.save_json = save_json
-        self.cur_local_message_round = 0
-        self.phase_init_bots()
 
+class GamePlayAsync(GamePlay):
+    @gen.coroutine  
     def play(self):
         """play a game with the bots"""
 
         while not self.game.is_game_done:
-            self.step()
+            yield self.step()
 
         if self.save_json:
             to_saved_game_format(self.game, output_path='GamePlayFramework.json')
-
-    def phase_init_bots(self):
-        self.cur_local_message_round = 0
-        # reset bot round info
-        for bot in self.bots:
-            if type(bot) == BaselineMsgRoundBot:
-                bot.phase_init()
-
+        
     @gen.coroutine     
     def step(self):
         """one step of messaging"""
@@ -78,7 +51,7 @@ class GamePlay():
             rcvd_messages = list(rcvd_messages.items())
             
             # get messages to be sent from bot
-            ret_dict = bot(rcvd_messages)
+            ret_dict = yield bot(rcvd_messages)
 
 
             if "messages" in ret_dict:
@@ -109,3 +82,15 @@ class GamePlay():
 
         self.game.process()
         return {"messages": msgs_to_send}, self.game.is_game_done
+
+
+@gen.coroutine            
+def game_loop():
+    game_play_obj = GamePlayAsync(None, [RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot, RandomProposerBot_AsyncBot], 3, True)
+    yield game_play_obj.play()
+    stop_io_loop()
+                
+if __name__ == "__main__":
+    # from utils import OrdersData, MessagesData, get_order_tokens
+
+    start_io_loop(game_loop)

--- a/tests/bot_tests/gameplay_framework_async.py
+++ b/tests/bot_tests/gameplay_framework_async.py
@@ -31,7 +31,7 @@ class GamePlayAsync(GamePlay):
         """one step of messaging"""
 
         if self.game.is_game_done:
-            return True
+            return None, True
         
         # if message rounds over
         if self.cur_local_message_round == self.msg_rounds:
@@ -39,7 +39,7 @@ class GamePlayAsync(GamePlay):
         while self.game.get_current_phase()[-1] != 'M':
             self.game.process()
             if self.game.is_game_done:
-                return True
+                return None, True
 
         round_msgs = self.game.messages
         msgs_to_send = {}

--- a/tests/utils_test.py
+++ b/tests/utils_test.py
@@ -3,7 +3,7 @@ from baseline_bots.parsing_utils import dipnet_to_daide_parsing, daide_to_dipnet
 from diplomacy import Game, Message
 
 class TestUtils:
-    def test(self):
+    def test_get_list_of_orders(self):
         EXAMPLE_ORDER = 'A VIE S A BUD - GAL'
         EXAMPLE_ORDER_2 = 'A VIE H'
 
@@ -20,7 +20,7 @@ class TestUtils:
         orders_data.add_order(EXAMPLE_ORDER_2, overwrite=True)
         assert orders_data.get_list_of_orders() == ['A VIE H']
 
-
+    def test_sort_messages_by_most_recent(self):
         # test sort_messages_by_most_recent
         game = Game()
         powers = list(game.powers)
@@ -44,6 +44,7 @@ class TestUtils:
 
         assert sort_messages_by_most_recent(msgs)[0].message == "HELLO"
 
+    def test_dipnet_to_daide_parsing(self):
         # Tests for utils.dipnet_to_daide_parsing
         PARSING_TEST_CASES = [
             (["A PAR H"], ["(FRA AMY PAR) HLD"], False),
@@ -70,8 +71,7 @@ class TestUtils:
             assert daide_to_dipnet_parsing(tc_op[0])[0] == comparison_tc_op, (daide_to_dipnet_parsing(tc_op[0]), comparison_tc_op)
         
 
-
-
+    def test_dipnet_to_daide_parsing_convoys(self):
         # Tests for convoy orders
         PARSING_CVY_TEST_CASES = [
             (["A TUN - SYR VIA", "F ION C A TUN - SYR", "F EAS C A TUN - SYR"], ["(ITA AMY TUN) CTO SYR VIA (ION EAS)", "(ITA FLT ION) CVY (ITA AMY TUN) CTO SYR", "(ITA FLT EAS) CVY (ITA AMY TUN) CTO SYR"]),
@@ -87,9 +87,7 @@ class TestUtils:
                 assert daide_to_dipnet_parsing(tc_op_ord)[0] == tc_ip_ord.replace(" R ", " - "), (daide_to_dipnet_parsing(tc_op_ord), tc_ip_ord.replace(" R ", " - "))
         
 
-
-
-
+    def test_parse_proposal_messages(self):
         # Tests for parse_proposal_messages
         PARSE_PROPOSALS_TC = [
             [
@@ -217,9 +215,7 @@ class TestUtils:
                     assert set(parsed_orders_dict[pod_key][key]) == set(tc_op[pod_key][key]), (pod_key, key, set(parsed_orders_dict[pod_key][key]), set(tc_op[pod_key][key]))
 
 
-
-
-
+    def test_parse_FCT(self):
         # Tests for orders extraction
         FCT_TCS = [
             [
@@ -234,6 +230,7 @@ class TestUtils:
         for tc_ip, tc_op in FCT_TCS:
             assert parse_FCT(tc_ip) == tc_op, parse_FCT(tc_ip)
             
+    def test_parse_PRP(self):
         PRP_TCS = [
             [
                 "PRP (XDO (F BLK - CON))", 
@@ -247,6 +244,7 @@ class TestUtils:
         for tc_ip, tc_op in PRP_TCS:
             assert parse_PRP(tc_ip) == tc_op, parse_PRP(tc_ip)
 
+    def test_parse_arrangement(self):
         ORR_TCS = [
             [
                 "XDO (F BLK - CON)",
@@ -311,6 +309,8 @@ class TestUtils:
         for tc_ip, tc_op in ORR_XDO_ALY_TCS:
             assert parse_arrangement(tc_ip, xdo_only=False) == tc_op, (parse_arrangement(tc_ip, xdo_only=False), tc_op)
 
+
+    def test_smart_select_support_proposals(self):
         SMART_SELECT_SUPPORT_PROPOSALS = [
             [
                 {
@@ -332,6 +332,7 @@ class TestUtils:
             assert smart_select_support_proposals(tc_ip) == tc_op
 
         
+    def test_get_order_tokens(self):
         GET_ORDER_TOKENS_TCS = [
             [
                 "A PAR S A MAR - BUR",

